src/runtime.js

```diff
@@ -827,6 +828,25 @@ export class MultidocManager {
     }
   }
 
+  /**
+   * Find all incomplete images and unset their srcs, freeing up network
+   * resources.
+   * @param {!./service/ampdoc-impl.AmpDoc} ampdoc
+   * @private
+   */
+  haltLoadingImages_(ampdoc) {
+    const ampBody = ampdoc.getBody();
+    const imgs = ampBody && ampBody.getElementsByTagName('img');
+    for (let i = 0; imgs && i < imgs.length; i++) {
+      if (!imgs[i].complete) {
+        // Opt for tiny dataURI image instead of empty src to prevent the
+        // viewer from detecting a load error.
+        imgs[i].src =
+          'data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACwAAAAAAQABAAACAkQBADs=';
```

Comment:
We already send the `unload` signal to all elements when the document becomes inactive, which means this just needs to live as a `unlayoutCallback` on `amp-img`.

Reply:
Ah, thanks for the suggestion! I'll migrate this `src` update there later today.

Reply:
This turns out to be less than trivial. When `shadowDoc.close()` is called, I agree that `unlayout` callbacks on custom elements do run. Unfortunately, it's not super fast. Due to issue #9022, the general practice I use is:
```
// Close shadow AMP
shadowDoc.close();

// Restore host element
var hostElement = document.querySelector('#shadowHost');
document.body.removeChild(hostElement);
hostElement = document.createElement('div');
hostElement.id = 'shadowHost';
document.body.insertBefore(hostElement, document.body.firstChild);
```
Removal of the shadowRoot host element right after calling `shadowDoc.close()` interrupts the unlayout. I've verified this be adding a timeout before removing `hostElement`.

I don't want to set an arbitrary timeout on `hostElement` removal. I'm also unsure how important it is to run all `unlayout` callbacks. Is there a message broadcast when unlayout finishes so I could start timing it on different devices and decide whether it's worth the wait?

Reply:
Commit 3eae496 in this PR adopts your suggestion to move the image interruption to `amp-img` `unlayout`. By itself, it is not yet effective for the example PWA included here. PR #25094 adds a `.closeAsync()` shadowDoc method (so that we don't change the API for those using `.close()` already) which can be awaited to ensure cleanup is complete, including the `unlayout` step this PR.